### ISSUE TITLE
data_layer.cpp:148] Check failed: false

### ISSUE THREAD

**Issue by cliffdean** [NVIDIA/caffe#384]
Hi,
I used NVCaffe 0.16.3 to show the following error

I0803 14:06:56.940691 16017 base_data_layer.cpp:217] [0] Prefetch allocated.
I0803 14:06:56.941608 16017 internal_thread.cpp:40] Restarting 4 internal thread(s) on device 0
I0803 14:06:56.941833 16017 data_reader.cpp:52] Data Reader threads: 4, out queues: 16, depth: 5
I0803 14:06:56.941890 16017 internal_thread.cpp:19] Starting 4 internal thread(s) on device 0
I0803 14:06:56.942250 16015 base_data_layer.cpp:217] [0] Prefetch allocated.
I0803 14:06:56.943159 16015 internal_thread.cpp:40] Restarting 4 internal thread(s) on device 0
I0803 14:06:56.943295 16015 data_reader.cpp:52] Data Reader threads: 4, out queues: 16, depth: 5
I0803 14:06:56.943354 16015 internal_thread.cpp:19] Starting 4 internal thread(s) on device 0
I0803 14:06:56.944249 16036 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/features
I0803 14:06:56.960284 16038 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/features
I0803 14:06:56.961181 16037 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/features
I0803 14:06:56.962069 16027 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/labels
I0803 14:06:56.962118 16017 data_layer.cpp:184] [0] ReshapePrefetch 5, 1, 25, 16
I0803 14:06:56.962798 16017 data_layer.cpp:208] [0] Output data size: 5, 1, 25, 16
I0803 14:06:56.963598 16017 base_data_layer.cpp:217] [0] Prefetch allocated.
I0803 14:06:56.963608 16017 data_layer.cpp:97] [0] Parser threads: 4 (auto)
I0803 14:06:56.963613 16017 data_layer.cpp:99] [0] Transformer threads: 4 (auto)
I0803 14:06:56.964143 16028 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/labels
I0803 14:06:56.964601 16029 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/labels
I0803 14:06:56.965056 16030 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/labels
I0803 14:06:56.965518 16035 db_lmdb.cpp:35] Opened lmdb /var/smb/work/digits/digits/jobs/20170801-084003-e62d/train_db/features
I0803 14:06:56.978515 16015 data_layer.cpp:184] [0] ReshapePrefetch 5, 3, 384, 1248
I0803 14:06:56.978966 16015 data_layer.cpp:208] [0] Output data size: 5, 3, 384, 1248
I0803 14:06:56.986868 16015 base_data_layer.cpp:217] [0] Prefetch allocated.
I0803 14:06:56.987268 16031 base_data_layer.cpp:217] [0] Prefetch allocated.
I0803 14:06:56.988025 16031 internal_thread.cpp:40] Restarting 4 internal thread(s) on device 0
F0803 14:06:56.988338 16031 data_layer.cpp:148] Check failed: false
*** Check failure stack trace: ***
@     0x7f60553df5cd  google::LogMessage::Fail()
@     0x7f60553e1433  google::LogMessage::SendToLog()
@     0x7f60553df15b  google::LogMessage::Flush()
@     0x7f60553e1e1e  google::LogMessageFatal::~LogMessageFatal()
@     0x7f6055e667bd  caffe::DataLayer<>::DataLayerSetUp()
@     0x7f6055d95a8b  caffe::BasePrefetchingDataLayer<>::InitializePrefetch()
@     0x7f6055e6c856  caffe::DataLayer<>::InitializePrefetch()
@     0x7f6055d8dd8f  caffe::BasePrefetchingDataLayer<>::InternalThreadEntryN()
@     0x7f6055c4c2aa  caffe::InternalThread::entry()
I0803 14:06:56.988760 16015 data_layer.cpp:97] [0] Parser threads: 4 (auto)
I0803 14:06:56.999456 16015 data_layer.cpp:99] [0] Transformer threads: 4 (auto)
@     0x7f6055c4e33b  boost::detail::thread_data<>::run()
@     0x7f60487e65d5  (unknown)
@     0x7f603b7606ba  start_thread
@     0x7f6053c5c3dd  clone
@              (nil)  (unknown)
 
The "threads_.size" error occurred with the previous version ,see  #363 
What should I do to solve it?
thanks

**Comment by drnikolaev** [comment]
Hi @cliffdean , may I have the complete log attached please?
When I run the model attached to #363 it works perfectly fine:

```
I0804 13:46:14.398257 14813 base_data_layer.cpp:217] [1] Prefetch allocated.
I0804 13:46:14.417969 14806 base_data_layer.cpp:217] [0] Prefetch allocated.
I0804 13:46:14.423125 14806 internal_thread.cpp:40] Restarting 4 internal thread(s) on device 0
I0804 13:46:14.423470 14806 data_reader.cpp:52] Data Reader threads: 4, out queues: 16, depth: 32
I0804 13:46:14.423598 14806 internal_thread.cpp:19] Starting 4 internal thread(s) on device 0
I0804 13:46:14.423696 14813 internal_thread.cpp:40] Restarting 4 internal thread(s) on device 1
I0804 13:46:14.424293 14813 data_reader.cpp:52] Data Reader threads: 4, out queues: 16, depth: 32
I0804 13:46:14.424576 14813 internal_thread.cpp:19] Starting 4 internal thread(s) on device 1
I0804 13:46:14.424796 14824 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.424909 14806 data_layer.cpp:184] [0] ReshapePrefetch 32, 3, 32, 32
I0804 13:46:14.425815 14822 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.426380 14806 data_layer.cpp:208] [0] Output data size: 32, 3, 32, 32
I0804 13:46:14.426818 14823 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.427690 14826 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.427852 14806 base_data_layer.cpp:217] [0] Prefetch allocated.
I0804 13:46:14.427870 14806 data_layer.cpp:97] [0] Parser threads: 4 (auto)
I0804 13:46:14.427876 14806 data_layer.cpp:99] [0] Transformer threads: 4 (auto)
I0804 13:46:14.427932 14810 solver.cpp:316] Iteration 0 (0.144039 s), loss = 2.32056
I0804 13:46:14.427959 14810 solver.cpp:333]     Train net output #0: loss = 2.32056 (* 1 = 2.32056 loss)
I0804 13:46:14.427971 14810 sgd_solver.cpp:140] Iteration 0, lr = 0.1, m = 0.9
I0804 13:46:14.428719 14830 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.428786 14813 data_layer.cpp:184] [1] ReshapePrefetch 32, 3, 32, 32
I0804 13:46:14.429471 14831 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.430614 14832 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.430812 14813 data_layer.cpp:208] [1] Output data size: 32, 3, 32, 32
I0804 13:46:14.431440 14833 db_lmdb.cpp:35] Opened lmdb /home/snikolaev/CODE/caffe/drnikolaev/caffe/examples/cifar10/cifar10_train_lmdb
I0804 13:46:14.432598 14813 base_data_layer.cpp:217] [1] Prefetch allocated.
I0804 13:46:14.432621 14813 data_layer.cpp:97] [1] Parser threads: 4 (auto)
I0804 13:46:14.432627 14813 data_layer.cpp:99] [1] Transformer threads: 4 (auto)
I0804 13:46:14.439803 14818 blocking_queue.cpp:40] Waiting for datum
I0804 13:46:14.471515 14810 solver.cpp:316] Iteration 1 (0.0435836 s), loss = 1.95111
I0804 13:46:14.471541 14810 solver.cpp:333]     Train net output #0: loss = 1.95111 (* 1 = 1.95111 loss)
I0804 13:46:14.480587 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'conv1a' with space 0.68G/1 1 0 0  (limit 15.07G, req 0.02G)
I0804 13:46:14.481062 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'conv1a' with space 0.68G/1 1 0 0  (limit 15.32G, req 0.02G)
I0804 13:46:14.514688 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'conv1b' with space 1.35G/2 1 1 0  (limit 14.4G, req 0.02G)
I0804 13:46:14.520611 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'conv1b' with space 1.35G/2 1 1 3  (limit 14.64G, req 0.02G)
I0804 13:46:14.526510 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res2a_branch2a' with space 1.35G/1 6 4 3  (limit 14.4G, req 0.02G)
I0804 13:46:14.532613 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res2a_branch2a' with space 1.35G/1 6 4 3  (limit 14.64G, req 0.02G)
I0804 13:46:14.533252 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res2a_branch2b' with space 1.35G/2 1 4 0  (limit 14.4G, req 0.02G)
I0804 13:46:14.539284 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res2a_branch2b' with space 1.35G/2 1 4 0  (limit 14.64G, req 0.02G)
I0804 13:46:14.547654 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res3a_branch2a' with space 1.35G/1 6 5 5  (limit 14.4G, req 0.02G)
I0804 13:46:14.549674 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res3a_branch2a' with space 1.35G/1 6 4 5  (limit 14.64G, req 0.02G)
I0804 13:46:14.553769 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res3a_branch2b' with space 1.35G/2 6 4 0  (limit 14.4G, req 0.02G)
I0804 13:46:14.554430 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res3a_branch2b' with space 1.35G/2 6 4 0  (limit 14.64G, req 0.02G)
I0804 13:46:14.576303 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res4a_branch2a' with space 1.35G/1 7 5 5  (limit 14.64G, req 0.03G)
I0804 13:46:14.578711 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res4a_branch2a' with space 1.35G/1 7 5 5  (limit 14.4G, req 0.03G)
I0804 13:46:14.585383 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res4a_branch2b' with space 1.35G/2 6 4 3  (limit 14.64G, req 0.03G)
I0804 13:46:14.586199 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res4a_branch2b' with space 1.35G/2 6 4 3  (limit 14.4G, req 0.03G)
I0804 13:46:14.614783 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res5a_branch2a' with space 1.35G/1 7 5 5  (limit 14.64G, req 0.03G)
I0804 13:46:14.617123 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res5a_branch2a' with space 1.35G/1 7 5 5  (limit 14.4G, req 0.03G)
I0804 13:46:14.624138 14811 cudnn_conv_layer.cpp:1014] [1] Conv Algos (F,BD,BF): 'res5a_branch2b' with space 1.35G/2 7 5 5  (limit 14.64G, req 0.03G)
I0804 13:46:14.625710 14810 cudnn_conv_layer.cpp:1014] [0] Conv Algos (F,BD,BF): 'res5a_branch2b' with space 1.35G/2 7 5 5  (limit 14.4G, req 0.03G)
I0804 13:46:14.635905 14810 solver.cpp:316] Iteration 2 (0.164382 s), loss = 2.35239
I0804 13:46:14.635931 14810 solver.cpp:333]     Train net output #0: loss = 2.35239 (* 1 = 2.35239 loss)
I0804 13:46:14.635962 14811 cudnn_conv_layer.cpp:294] [1] Layer 'conv1a' reallocating workspace: 1.35G -> 0.07G
I0804 13:46:14.635982 14810 cudnn_conv_layer.cpp:294] [0] Layer 'conv1a' reallocating workspace: 1.35G -> 0.07G
I0804 13:46:15.964927 14810 solver.cpp:311] Iteration 100 (73.7407 iter/s, 1.32898s/98 iter), loss = 1.56079
I0804 13:46:15.964951 14810 solver.cpp:333]     Train net output #0: loss = 1.56079 (* 1 = 1.56079 loss)
I0804 13:46:15.964956 14810 sgd_solver.cpp:140] Iteration 100, lr = 0.0998438, m = 0.9
I0804 13:46:17.279551 14810 solver.cpp:311] Iteration 200 (76.0699 iter/s, 1.31458s/100 iter), loss = 1.59147
I0804 13:46:17.279574 14810 solver.cpp:333]     Train net output #0: loss = 1.59147 (* 1 = 1.59147 loss)
I0804 13:46:17.279579 14810 sgd_solver.cpp:140] Iteration 200, lr = 0.0996875, m = 0.9
I0804 13:46:18.604789 14810 solver.cpp:311] Iteration 300 (75.4606 iter/s, 1.3252s/100 iter), loss = 1.6555
I0804 13:46:18.604815 14810 solver.cpp:333]     Train net output #0: loss = 1.6555 (* 1 = 1.6555 loss)
I0804 13:46:18.604820 14810 sgd_solver.cpp:140] Iteration 300, lr = 0.0995313, m = 0.9
I0804 13:46:19.926281 14810 solver.cpp:311] Iteration 400 (75.6746 iter/s, 1.32145s/100 iter), loss = 1.25313
I0804 13:46:19.926308 14810 solver.cpp:333]     Train net output #0: loss = 1.25313 (* 1 = 1.25313 loss)
I0804 13:46:19.926316 14810 sgd_solver.cpp:140] Iteration 400, lr = 0.099375, m = 0.9
^CI0804 13:46:20.051756 14810 solver.cpp:589] Snapshotting to binary proto file jacintonet11v2_cifar10_iter_409.caffemodel
I0804 13:46:20.080476 14810 sgd_solver.cpp:367] Snapshotting solver state to binary proto file jacintonet11v2_cifar10_iter_409.solverstate
```

**Comment by drnikolaev** [comment]
Hi @cliffdean , I found and fixed a race condition in DataLayerSetUp. Most probably, it was the reason of all failures you observed. Please clone the latest code from the branch and give it a try. Feel free to reopen the issue if it still crashes. Thank you for reporting!
